engine/src/main/java/org/terasology/world/zones/LayeredZoneRegionFunction.java

```diff
@@ -0,0 +1,200 @@
+        this.minWidth = minWidth;
+        this.maxWidth = maxWidth;
+        this.ordering = ordering;
+    }
+
+    @Override
+    public boolean apply(int x, int y, int z, Region region, Zone zone) {
+        return getLayerRange(x, z, region, zone).layerContains(y);
+    }
+
+    private LayerRange getLayerRange(int x, int z, Region region, Zone zone) {
+        Vector2i pos = new Vector2i(x, z);
+        if (!layerRangeMap.containsKey(pos)) {
+            int surfaceHeight = (int) Math.floor(region.getFacet(SurfaceHeightFacet.class).getWorld(pos));
+
+            boolean aboveground = ordering >= 0;
+            int cumulativeDistanceSmall = 0;
+            int cumulativeDistanceLarge = 0;
+            LayerRange layerRange = null;
+
+            List<LayeredZoneRegionFunction> layers =
+                    aboveground ? getAbovegroundLayers(zone) : getUndergroundLayers(zone);
+
+            int i;
+            for (i = 0; i < layers.size(); i++) {
+                LayeredZoneRegionFunction layer = layers.get(i);
+                //TODO: allow variable-width layers
+                cumulativeDistanceLarge += layer.getMinWidth();
+                if (this.equals(layer)) {
+                    if (aboveground) {
+                        layerRange = new LayerRange()
+                                .setMin(surfaceHeight + cumulativeDistanceSmall)
+                                .setMax(surfaceHeight + cumulativeDistanceLarge);
+                        break;
+                    } else {
+                        layerRange = new LayerRange()
+                                .setMin(surfaceHeight -cumulativeDistanceLarge)
+                                .setMax(surfaceHeight - cumulativeDistanceSmall);
+                        break;
+                    }
+                }
+                cumulativeDistanceSmall += layer.getMinWidth();
+            }
+
+            if (layers.size() <= 0 || layerRange == null) {
+                throw new IllegalStateException("Layer for zone '" + zone + "' not found in list of " +
+                        (aboveground ? "aboveground" : "underground") + " layers.");
+            }
+
+            if (i == layers.size() - 1) {
+                //At one of the edge layers
+                if (aboveground) {
+                    layerRange.unsetMax();
+                } else {
+                    layerRange.unsetMin();
+                }
+            }
+            layerRangeMap.put(pos, layerRange);
+        }
+        return layerRangeMap.get(pos);
+    }
+
+    private List<LayeredZoneRegionFunction> getSiblings(Zone zone) {
+        if (siblings == null) {
+            siblings = getSiblingRegionFunctions(zone).stream()
+                    .filter(f -> f instanceof LayeredZoneRegionFunction)
+                    .map(l -> (LayeredZoneRegionFunction) l)
+                    .sorted((l1, l2) -> ((Integer) Math.abs(l1.getOrdering())).compareTo(Math.abs(l2.getOrdering())))
```

Comment:
it's layer1 and layer2 I guess, although it is confusing that you mix up `f` for function and `l` for layer here.